File: server/sync.ts
import { adminDb } from './firebase';
import { SheetsService } from './sheets';
import { FacebookProvider, ZaloOAProvider, MockProvider, SocialProvider } from './providers';
import { Channel, Post, DailySnapshot, ApiLog } from '../src/types';
import { v4 as uuidv4 } from 'uuid';

// In-memory sync locks
const syncLocks = new Set<string>();

export class SyncEngine {
  private static getProvider(platform: string): SocialProvider {
    switch (platform) {
      case 'facebook':
        return new FacebookProvider();
      case 'zalo':
        return new ZaloOAProvider();
      case 'mock':
        return new MockProvider();
      default:
        throw new Error(`Nền tảng "${platform}" chưa được hỗ trợ.`);
    }
  }

  /**
   * Đồng bộ hóa dữ liệu của một Kênh cụ thể
   */
  static async syncChannel(
    channelId: string,
    googleAccessToken: string | null,
    sinceDate?: Date,
    untilDate?: Date,
    requestId: string = uuidv4()
  ): Promise<any> {
    if (syncLocks.has(channelId)) {
      throw new Error('Kênh này đang được đồng bộ hóa. Vui lòng thử lại sau.');
    }

    syncLocks.add(channelId);
    const startedAt = new Date().toISOString();

    // 1. Đọc kênh từ Firestore
    const channelRef = adminDb.collection('channels').doc(channelId);
    const channelSnap = await channelRef.get();
    if (!channelSnap.exists) {
      syncLocks.delete(channelId);
      throw new Error(`Không tìm thấy kênh ID ${channelId}`);
    }

    const channel = channelSnap.data() as Channel;
    if (channel.status === 'inactive') {
      syncLocks.delete(channelId);
      throw new Error('Kênh đã bị vô hiệu hóa.');
    }

    const provider = this.getProvider(channel.platform);
    let logId = uuidv4();
    let recordsReceived = 0;
    let recordsInserted = 0;
    let recordsUpdated = 0;
    let status: 'success' | 'failed' = 'success';
    let errorCode = '';
    let errorMessage = '';

    try {
      // 2. Gọi API nhà mạng lấy danh sách bài đăng
      const rawPosts = await provider.listPosts(channelId, channel.externalId, sinceDate, untilDate);
      recordsReceived = rawPosts.length;

      const postsToUpsert: Post[] = [];
      const snapshotsToUpsert: DailySnapshot[] = [];

      if (rawPosts.length > 0) {
        // 3. Lấy metrics cho các bài đăng nhận được
        const rawMetrics = await provider.getPostMetrics(channelId, channel.externalId, rawPosts);

        // Múi giờ Asia/Bangkok cho ngày snapshot_date
        const tzOffset = 7 * 60 * 60 * 1000; // +7 hours
        const bkkTime = new Date(Date.now() + tzOffset);
        const snapshotDate = bkkTime.toISOString().split('T')[0]; // yyyy-MM-dd

        for (let i = 0; i < rawPosts.length; i++) {
          const rawPost = rawPosts[i];
          const rawMetric = rawMetrics.find(m => m.id === rawPost.id) || { id: rawPost.id };

          const normalizedPost = provider.normalizePost(rawPost, channelId);
          const normalizedSnapshot = provider.normalizeMetrics(rawMetric, channelId, normalizedPost.postKey, snapshotDate);

          postsToUpsert.push(normalizedPost);
          snapshotsToUpsert.push(normalizedSnapshot);
        }

        // 4. Upsert vào Firestore DB
        // Firestore batch upsert bài viết
        const batch = adminDb.batch();
        for (const post of postsToUpsert) {
          const postRef = adminDb.collection('posts').doc(post.postKey);
          batch.set(postRef, post, { merge: true });
        }
        await batch.commit();

        // Firestore batch upsert snapshot ngày
        const snapBatch = adminDb.batch();
        for (const snapshot of snapshotsToUpsert) {
          const snapRef = adminDb.collection('dailySnapshots').doc(snapshot.snapshotKey);
          snapBatch.set(snapRef, snapshot, { merge: true });
        }
        await snapBatch.commit();

        // Đếm số dòng thay đổi thực tế hoặc giả lập ghi nhận
        recordsInserted = postsToUpsert.length;
        recordsUpdated = snapshotsToUpsert.length;
      }

      // 5. Cập nhật thông tin kênh
      const updatedChannelFields = {
        lastSyncAt: new Date().toISOString(),
        lastSyncStatus: 'success',
        totalPosts: (channel.totalPosts || 0) + recordsInserted
      };
      await channelRef.update(updatedChannelFields);

      // 6. Ghi dữ liệu vào Google Sheets (Nếu có token và spreadsheetId)
      let spreadsheetId = '';
      const configSnap = await adminDb.collection('systemConfig').doc('main').get();
      if (configSnap.exists) {
        spreadsheetId = configSnap.data()?.spreadsheetId || '';
      }

      if (googleAccessToken && spreadsheetId) {
        try {
          const sheetsService = new SheetsService(googleAccessToken, spreadsheetId);
          
          // Upsert Kênh MXH dòng hiện tại
          const channelRow = {
            channel_id: channel.id,
            platform: channel.platform,
            channel_name: channel.name,
            external_id: channel.externalId,
            channel_url: channel.platform === 'facebook' ? `https://facebook.com/${channel.externalId}` : `https://oa.zalo.me/${channel.externalId}`,
            status: channel.status,
            timezone: channel.timezone || 'Asia/Bangkok',
            created_at: channel.createdAt,
            updated_at: new Date().toISOString(),
            last_sync_at: updatedChannelFields.lastSyncAt,
            last_sync_status: updatedChannelFields.lastSyncStatus,
            total_posts: updatedChannelFields.totalPosts
          };
          await sheetsService.upsertRecords('KENH_MXH', 'channel_id', [channelRow]);

          // Upsert Bài viết
          if (postsToUpsert.length > 0) {
            const sheetPosts = postsToUpsert.map(p => ({
              post_key: p.postKey,
              platform: p.platform,
              channel_id: p.channelId,
              external_post_id: p.externalPostId,
              post_url: p.postUrl,
              post_type: p.postType,
              message: p.message,
              published_at: p.publishedAt,
              imported_at: p.importedAt,
              updated_at: p.updatedAt,
              is_deleted: p.isDeleted
            }));
            await sheetsService.upsertRecords('BAI_DANG', 'post_key', sheetPosts);
          }

          // Upsert snapshot dữ liệu ngày
          if (snapshotsToUpsert.length > 0) {
            const sheetSnapshots = snapshotsToUpsert.map(s => ({
              snapshot_key: s.snapshotKey,
              snapshot_date: s.snapshotDate,
              platform: s.platform,
              channel_id: s.channelId,
              post_key: s.postKey,
              reactions: s.reactions,
              likes: s.likes,
              comments: s.comments,
              shares: s.shares,
              views: s.views,
              reach: s.reach,
              impressions: s.impressions,
              clicks: s.clicks,
              total_engagement: s.totalEngagement,
              engagement_rate: s.engagementRate,
              fetched_at: s.fetchedAt
            }));
            await sheetsService.upsertRecords('DU_LIEU_NGAY', 'snapshot_key', sheetSnapshots);
          }
        } catch (sheetError: any) {
          console.error('Lỗi khi đồng bộ Google Sheets:', sheetError.message);
          // Không quăng lỗi ra ngoài làm hỏng toàn bộ luồng đồng bộ API chính
        }
      }

    } catch (error: any) {
      status = 'failed';
      errorCode = error.code || 'SYNC_ERROR';
      errorMessage = error.message || 'Lỗi đồng bộ hóa chưa xác định';
      console.error(`Lỗi đồng bộ hóa kênh ${channelId}:`, error);

      // Cập nhật trạng thái lỗi cho Kênh
      await channelRef.update({
        lastSyncAt: new Date().toISOString(),
        lastSyncStatus: 'failed'
      }).catch(console.error);

    } finally {
      const endedAt = new Date().toISOString();
      const apiLog: ApiLog = {
        logId,
        startedAt,
        endedAt,
        platform: channel.platform,
        action: 'sync',
        channelId,
        status: status === 'success' ? 'success' : 'failed',
        recordsReceived,
        recordsInserted,
        recordsUpdated,
        errorCode,
        errorMessage,
        requestId
      };

      // Lưu Log vào Firestore
      await adminDb.collection('apiLogs').doc(logId).set(apiLog).catch(console.error);

      // Ghi log vào Google Sheets nếu có thể
      if (googleAccessToken) {
        try {
          const configSnap = await adminDb.collection('systemConfig').doc('main').get();
          const spreadsheetId = configSnap.data()?.spreadsheetId;
          if (spreadsheetId) {
            const sheetsService = new SheetsService(googleAccessToken, spreadsheetId);
            await sheetsService.upsertRecords('NHAT_KY_API', 'log_id', [{
              log_id: apiLog.logId,
              started_at: apiLog.startedAt,
              ended_at: apiLog.endedAt,
              platform: apiLog.platform,
              action: apiLog.action,
              channel_id: apiLog.channelId,
              status: apiLog.status,
              records_received: apiLog.recordsReceived,
              records_inserted: apiLog.recordsInserted,
              records_updated: apiLog.recordsUpdated,
              error_code: apiLog.errorCode || '',
              error_message: apiLog.errorMessage || '',
              request_id: apiLog.requestId
            }]);
          }
        } catch (e: any) {
          console.error('Không thể ghi nhật ký API vào Google Sheet:', e.message);
        }
      }

      syncLocks.delete(channelId);
    }

    if (status === 'failed') {
      throw new Error(errorMessage);
    }

    return {
      success: true,
      recordsReceived,
      recordsInserted,
      recordsUpdated,
    };
  }

  /**
   * Đồng bộ hóa tất cả các kênh đang hoạt động
   */
  static async syncAllChannels(googleAccessToken: string | null, requestId: string = uuidv4()): Promise<any> {
    const channelsSnap = await adminDb.collection('channels').where('status', '==', 'active').get();
    const results: any[] = [];
    
    for (const doc of channelsSnap.docs) {
      const channelId = doc.id;
      try {
        const res = await this.syncChannel(channelId, googleAccessToken, undefined, undefined, requestId);
        results.push({ channelId, success: true, ...res });
      } catch (e: any) {
        results.push({ channelId, success: false, error: e.message });
      }
    }
    return results;
  }
}

File: backend/social/sync.py
import datetime
import uuid

from django.utils import timezone
from django.utils.dateparse import parse_date

from .models import ApiLog, Channel, DailySnapshot, FollowerSnapshot, Post
from .providers import FacebookProvider, MockProvider, ZaloOAProvider


DEFAULT_SYNC_DAYS = 365


class SyncEngine:
    @staticmethod
    def queue_channels(channels, request_id):
        queued_at = timezone.now()
        queued_logs = {}
        for channel in channels:
            log = ApiLog.objects.create(
                log_id=f"queue_{int(queued_at.timestamp() * 1000)}_{uuid.uuid4().hex[:5]}",
                started_at=queued_at,
                platform=channel.platform,
                action="sync_channel",
                channel_id=channel.id,
                status="queued",
                request_id=request_id,
            )
            queued_logs[channel.id] = log
        return queued_logs

    @staticmethod
    def get_provider(platform):
        platform = platform.lower()
        if platform == "facebook":
            return FacebookProvider()
        if platform == "zalo":
            return ZaloOAProvider()
        return MockProvider()

    @staticmethod
    def _parse_boundary(value, end_of_day=False):
        if not value or isinstance(value, datetime.datetime):
            return value
        if isinstance(value, datetime.date):
            parsed = value
        else:
            parsed = parse_date(str(value))
        if not parsed:
            return value

        boundary_time = datetime.time.max if end_of_day else datetime.time.min
        result = datetime.datetime.combine(parsed, boundary_time)
        return timezone.make_aware(result, timezone.get_current_timezone())

    @staticmethod
    def _save_snapshot(provider, channel, post_key, raw_metric, snapshot_date):
        metric_data = provider.normalize_metrics(raw_metric, channel.id, post_key, snapshot_date)
        DailySnapshot.objects.update_or_create(
            snapshot_key=metric_data["snapshot_key"],
            defaults={
                "snapshot_date": metric_data["snapshot_date"],
                "platform": metric_data["platform"],
                "channel_id": metric_data["channel_id"],
                "post_key": metric_data["post_key"],
                "reactions": metric_data["reactions"],
                "likes": metric_data["likes"],
                "comments": metric_data["comments"],
                "shares": metric_data["shares"],
                "views": metric_data["views"],
                "reach": metric_data["reach"],
                "impressions": metric_data["impressions"],
                "clicks": metric_data["clicks"],
                "total_engagement": metric_data["total_engagement"],
                "engagement_rate": metric_data["engagement_rate"],
                "fetched_at": metric_data["fetched_at"],
            },
        )

    @classmethod
    def sync_channel(
        cls,
        channel_id,
        request_id=None,
        since=None,
        until=None,
        snapshot_existing_since=None,
        queued_log=None,
    ):
        request_id = request_id or f"req_{uuid.uuid4().hex[:10]}"
        started_at = timezone.now()
        log_id = f"sync_{int(started_at.timestamp() * 1000)}_{uuid.uuid4().hex[:5]}"

        since = cls._parse_boundary(since)
        until = cls._parse_boundary(until, end_of_day=True)
        snapshot_existing_since = cls._parse_boundary(snapshot_existing_since)

        if not since:
            since = timezone.now() - datetime.timedelta(days=DEFAULT_SYNC_DAYS)
        if not until:
            until = timezone.now()

        try:
            channel = Channel.objects.get(id=channel_id)
        except Channel.DoesNotExist:
            if queued_log:
                queued_log.status = "failed"
                queued_log.ended_at = timezone.now()
                queued_log.error_code = "CHANNEL_NOT_FOUND"
                queued_log.error_message = "Channel not found"
                queued_log.save(
                    update_fields=[
                        "status",
                        "ended_at",
                        "error_code",
                        "error_message",
                    ]
                )
            else:
                ApiLog.objects.create(
                    log_id=log_id,
                    started_at=started_at,
                    ended_at=timezone.now(),
                    platform="unknown",
                    action="sync_channel",
                    channel_id=channel_id,
                    status="failed",
                    error_code="CHANNEL_NOT_FOUND",
                    error_message="Channel not found",
                    request_id=request_id,
                )
            return False, "Không tìm thấy kênh"

        if queued_log:
            api_log = queued_log
            api_log.started_at = started_at
            api_log.platform = channel.platform
            api_log.status = "running"
            api_log.ended_at = None
            api_log.error_code = None
            api_log.error_message = None
            api_log.save(
                update_fields=[
                    "started_at",
                    "platform",
                    "status",
                    "ended_at",
                    "error_code",
                    "error_message",
                ]
            )
        else:
            api_log = ApiLog.objects.create(
                log_id=log_id,
                started_at=started_at,
                platform=channel.platform,
                action="sync_channel",
                channel_id=channel_id,
                status="running",
                request_id=request_id,
            )

        try:
            provider = cls.get_provider(channel.platform)
            snapshot_date = timezone.localdate().isoformat()

            followers = provider.get_followers(channel.id, channel.external_id)
            raw_follower_insights = getattr(provider, "get_follower_insights", lambda *_args, **_kwargs: [])(
                channel.id,
                channel.external_id,
                since=since,
                until=until,
            )
            follower_insights = raw_follower_insights if isinstance(raw_follower_insights, list) else []
            for insight in follower_insights:
                if not isinstance(insight, dict):
                    continue
                insight_date = str(insight.get("snapshot_date") or "").strip()
                if not insight_date or "followers_count" not in insight:
                    continue
                defaults = {
                    "snapshot_date": insight_date,
                    "channel_id": channel.id,
                    "channel_name": channel.name,
                    "followers_count": insight["followers_count"],
                    "fetched_at": timezone.now(),
                }
                for field in ("daily_follows_unique", "daily_unfollows_unique"):
                    if field in insight:
                        defaults[field] = insight[field]
                FollowerSnapshot.objects.update_or_create(
                    snapshot_key=f"{insight_date}:{channel.id}",
                    defaults=defaults,
                )

            # The regular Page field is the freshest stock count. Do not overwrite
            # same-day daily Insights values that may already have been saved above.
            FollowerSnapshot.objects.update_or_create(
                snapshot_key=f"{snapshot_date}:{channel.id}",
                defaults={
                    "snapshot_date": snapshot_date,
                    "channel_id": channel.id,
                    "channel_name": channel.name,
                    "followers_count": followers,
                    "fetched_at": timezone.now(),
                },
            )

            raw_posts = provider.list_posts(
                channel.id,
                channel.external_id,
                since=since,
                until=until,
            )
            raw_posts = [post for post in raw_posts if post.get("id")]
            api_log.records_received = len(raw_posts)

            inserted = 0
            updated = 0
            post_keys_by_external_id = {}

            for raw_post in raw_posts:
                post_data = provider.normalize_post(raw_post, channel.id)
                post_obj, created = Post.objects.update_or_create(
                    post_key=post_data["post_key"],
                    defaults={
                        "platform": post_data["platform"],
                        "channel_id": post_data["channel_id"],
                        "external_post_id": post_data["external_post_id"],
                        "post_url": post_data["post_url"],
                        "image_url": post_data.get("image_url"),
                        "post_type": post_data["post_type"],
                        "message": post_data.get("message"),
                        "published_at": post_data["published_at"],
                        "imported_at": post_data["imported_at"],
                        "updated_at": post_data["updated_at"],
                        "is_deleted": post_data["is_deleted"],
                    },
                )
                post_keys_by_external_id[raw_post["id"]] = post_obj.post_key
                if created:
                    inserted += 1
                else:
                    updated += 1

            metric_targets = list(raw_posts)
            if snapshot_existing_since:
                known_external_ids = set(post_keys_by_external_id)
                tracked_posts = Post.objects.filter(
                    channel_id=channel.id,
                    is_deleted=False,
                    published_at__gte=snapshot_existing_since,
                    published_at__lte=until,
                ).only("post_key", "external_post_id")

                for tracked_post in tracked_posts.iterator():
                    if tracked_post.external_post_id in known_external_ids:
                        continue
                    known_external_ids.add(tracked_post.external_post_id)
                    post_keys_by_external_id[tracked_post.external_post_id] = tracked_post.post_key
                    metric_targets.append({"id": tracked_post.external_post_id})

            snapshots_saved = 0
            if metric_targets:
                metrics = provider.get_post_metrics(
                    channel.id,
                    channel.external_id,
                    metric_targets,
                )
                for raw_metric in metrics:
                    external_id = raw_metric.get("id")
                    post_key = post_keys_by_external_id.get(external_id)
                    if not post_key:
                        continue
                    cls._save_snapshot(
                        provider,
                        channel,
                        post_key,
                        raw_metric,
                        snapshot_date,
                    )
                    snapshots_saved += 1

            channel.followers_count = followers
            channel.total_posts = Post.objects.filter(
                channel_id=channel.id,
                is_deleted=False,
            ).count()
            channel.last_sync_at = timezone.now()
            channel.last_sync_status = "success"
            channel.save(
                update_fields=[
                    "followers_count",
                    "total_posts",
                    "last_sync_at",
                    "last_sync_status",
                ]
            )

            api_log.status = "success"
            api_log.records_inserted = inserted
            api_log.records_updated = updated
            api_log.ended_at = timezone.now()
            api_log.save(
                update_fields=[
                    "status",
                    "records_received",
                    "records_inserted",
                    "records_updated",
                    "ended_at",
                ]
            )

            return (
                True,
                f"Thành công: thêm {inserted}, cập nhật {updated}, "
                f"lưu {snapshots_saved} snapshot ngày.",
            )

        except Exception as exc:
            channel.last_sync_status = "failed"
            channel.save(update_fields=["last_sync_status"])

            api_log.status = "failed"
            api_log.error_code = "SYNC_ERROR"
            api_log.error_message = str(exc)
            api_log.ended_at = timezone.now()
            api_log.save(
                update_fields=[
                    "status",
                    "error_code",
                    "error_message",
                    "ended_at",
                    "records_received",
                ]
            )
            return False, f"Lỗi đồng bộ: {exc}"

    @classmethod
    def sync_all_channels(
        cls,
        google_token=None,
        since=None,
        until=None,
        snapshot_existing_since=None,
    ):
        request_id = f"req_{uuid.uuid4().hex[:10]}"
        active_channels = list(Channel.objects.filter(status="active").order_by("name"))
        queued_logs = cls.queue_channels(active_channels, request_id)

        success_count = 0
        total_count = len(active_channels)
        results = []

        for channel in active_channels:
            success, message = cls.sync_channel(
                channel.id,
                request_id=request_id,
                since=since,
                until=until,
                snapshot_existing_since=snapshot_existing_since,
                queued_log=queued_logs[channel.id],
            )
            if success:
                success_count += 1
            results.append(
                {
                    "channel_id": channel.id,
                    "name": channel.name,
                    "success": success,
                    "message": message,
                }
            )

        return {
            "success": total_count > 0 and success_count == total_count,
            "message": (
                f"Đã đồng bộ thành công {success_count}/{total_count} "
                f"kênh đang hoạt động."
            ),
            "results": results,
            "requestId": request_id,
        }
